src/components/AddressHistory.js

```diff
@@ -73,27 +45,54 @@ class AddressHistory extends React.Component {
       return <p>This address does not have any transactions yet.</p>
     }
 
-    const getFirstHash = () => {
-      return this.props.transactions[0].tx_id;
+    const paginationLink = (page, query) => {
+        return this.props.pagination.setURLParameters({...query, page: page});
     }
 
-    const getLastHash = () => {
-      return this.props.transactions[this.props.transactions.length-1].tx_id;
-    }
+    const getPages = (page, total) => {
+      let start = page - 2;
+      let end = page + 2;
+      if (start < 0) {
+        // If the start is before 0, we push the end further
+        end -= start;
+        start = 0;
+      }
+
+      if (end > total) {
+        // If the start is not 0, push start back (up to 0)
+        start = start === 0 ? 0 : Math.max(start - end + total, 0);
+        end = total;
+      }
+
+      return Array(1 + end - start).fill().map((_, index) => start + index);
+    };
 
     const loadPagination = () => {
       if (this.props.transactions.length === 0) {
         return null;
       } else {
+        const queryParams = this.props.pagination.obtainQueryParams();
+        const page = +queryParams.page;
+        const lastPage = Math.ceil(this.props.numTransactions / TX_COUNT) - 1;
```

Comment:
Why `- 1`?

Reply:
An example:

187 tokens with page of length 10 would give us 19 pages (from 0 to 18).
The pages would be: page0 (token1 to token10), ...., page18 (token181 to token187).
But `ceil(187 / 10)` is 19, hence `ceil() - 1` is the last page.

This also works for exact numbers, e.g. 180 tokens would be 18 pages (from 0 to 17) this is why we don't use `floor` since it would give us one more empty page.

Reply:
Yeah, I guess the pagination starting at 0 made the confusion for me.